src/cleaner.ts

```diff
@@ -1,390 +1,535 @@
+			projectId: projectId,
+			tagsCount: true,
+		});
+		console.log(repos);
+	}
+
+	public async getGroupContainerRepositories(groupId: string | number) {
+		const repos = await this.gl.ContainerRegistry.allRepositories({
+			groupId: groupId,
+			tagsCount: true,
+		});
+		console.log(repos);
+	}
+
+	/**
+	 * Get all tags of a Project's Container Repository. Uses GitLab API pagination to run concurrent requests across multiple Promises,
+	 * each Promises having a range of pages to fetch.
+	 *
+	 * @param repository
+	 * @param tagsPerPage number of tags per page
+	 * @returns
+	 */
+	private async getRepositoryTagsConcurrently(
+		repository: RegistryRepositorySchema,
+		tagsPerPage = 50,
+	) {
+		const tagCount = repository.tags_count ?? 0;
+		const pageTotal = Math.ceil(tagCount / tagsPerPage);
+		const pages = [...Array(pageTotal).keys()].map((i) => i + 1);
+
+		console.log(
+			`🔭 Listing ${tagCount} tags (${pageTotal} pages, ${tagsPerPage} tags per page)`,
+		);
+
+		const result: CondensedRegistryRepositoryTagSchema[] = [];
+
+		// Process pages in batches of size this.concurrency
+		for (let i = 0; i < pages.length; i += this.concurrency) {
+			const batch = pages.slice(i, i + this.concurrency);
+			const batchPromises = batch.map((page) =>
+				this.getRepositoryTagsForPages(
+					repository.project_id,
+					repository.id,
+					page,
+					tagsPerPage,
+				),
+			);
+
+			const batchResults = await Promise.all(batchPromises);
+			result.push(...batchResults.flat());
+
+			console.log(`   Fetched ${result.length}/${tagCount} tags...`);
+		}
+
+		console.log(`   Found ${result.length} tags`);
+
+		return result;
+	}
+
+	/**
+	 * Fetch Container Repository tags for the given pages.
+	 */
+	private async getRepositoryTagsForPages(
+		projectId: number,
+		repositoryId: number,
+		page: number,
+		perPage: number,
+	) {
+		return await this.gl.ContainerRegistry.allTags(projectId, repositoryId, {
+			page: page,
+			perPage: perPage,
+		});
+	}
+
+	public async cleanupContainerRepositoryTags(
+		repositoryId: number,
+		options: Partial<CleanupContainerRepositoryTagsOptions> = {},
+	) {
+		const {
+			keepRegex = DEFAULT_KEEP_REGEX,
+			deleteRegex = DEFAULT_DELETE_REGEX,
+			olderThanDays = DEFAULT_OLDER_THAN_DAYS,
+			tagsPerPage = DEFAULT_TAGS_PER_PAGE,
+			outputTags = "",
+			keepMostRecentN = DEFAULT_KEEP_MOST_RECENT,
+		} = options;
+		console.log(
+			`🧹 Cleaning image tags for repository ${repositoryId}. Keep tags matching '${keepRegex}' and delete tags older than ${olderThanDays} days. Keeping ${keepMostRecentN} most recent tags. (dry-run: ${this.dryRun})`,
+		);
+
+		// warn user if parameters doesn't make sense or forgot to disable safety
+		if (
+			keepRegex === DEFAULT_KEEP_REGEX ||
+			deleteRegex === DEFAULT_DELETE_REGEX
+		) {
+			console.warn("");
+			console.warn(
+				`🤔 Hey, looks like you kept default keep and/or delete regex. By default, these regex won't mach anything for safety reasons.`,
```

Comment:
Coming back to this I wonder if making these flags required instead might not be a good idea ?

Reply:
I don't know, I think it's fine, as this requires the user to really think about what regexes they need, and is an additional safeguard.
I am not a big fan of "required options", if you understand what I mean.

Reply:
> this requires the user to really think about what regexes they need

Exactly what I had in mind initially. Let's keep it this way.